article/article.md

```diff
@@ -0,0 +1,153 @@
+
+Over time we grew dissatisfied with this approach, however. Not only was Carto challenging to set up and maintain, particularly in our [containerized microservice implementation](https://github.com/ihmeuw/cartodb-docker), but the tile server was quite slow at producing the colorized map tiles, making using the visualization tool a frustrating experience. We did some performance testing of the server with [Apache Jmeter](https://jmeter.apache.org/) and found that it was taking about 3 seconds, in the best case, to return a tile. I suspect we could have squeezed better performance out of the tile server with some optimizations, but its codebase was always something of an impenetrable black box to us, and the prospect of searching for needles in this haystack seemed quite daunting.
+
+## The Idea: Client-side rendering with WebGL
+
+Around this time, I began thinking about other ways to render our map tiles. One option was to create our own tile server to do the colorization. A custom server could be better suited to our specific needs, and, crucially, we'd be more able optimize its performance because we'd understand it better. Yet a server-side solution had some inherent limitations. Rendering a large number of pixels on the CPU is bound to be slow, because the pixels must be processed sequentially (or with limited parallelization taking advantage of multicore architecture). With commodity hardware and the VMs in which our containerized services run, there would be hard limits on how fast the tiles could be produced. It seemed to me that the ideal way to render the images was on the [GPU](https://en.wikipedia.org/wiki/Graphics_processing_unit), which is specifically designed to process large numbers of pixels in parallel. Fortunately, modern browsers provide a way to do GPU processing: [WebGL](https://developer.mozilla.org/en-US/docs/Web/API/WebGL_API)! A WebGL solution would mean the tile server would have very little to do - just retrieve pixel data from PostGIS and then pass it in essentially raw form to the client. The user's browser would then be responsible for actually colorizing the pixels, and it could take advantage of GPU rendering to do this very efficiently.
```

Comment:
I would recommend avoiding contractions "we'd" -> "we would " in the first sentence in two places.

I would also just write out Virtual Machines. I'm sure most people know what VM's are, but you never know.  I might also reword the sentence that start "With commodity hardware..." to be something like: "Because our applications are deployed on commodity hardware running containerized services in Virtual Machines, there would be hard limits on how fast the tiles could be produced."

Reply:
Revised this part to better communicate the hardware limitations of a server-side approach

Reply:
I eliminated "we'd." Since there are plenty of other contractions in the article, I'm curious if you'd propose any general principles on the use of contractions.

Reply:
I would suggest not using contractions in general. I like the writing style here for a technical article, but I think it will come across as a bit more professional without contractions.